psydac/linalg/tests/test_stencil_vector.py

```diff
@@ -745,6 +745,55 @@ def test_stencil_vector_2d_parallel_toarray(dtype, n1, n2, p1, p2, s1, s2, P1=Tr
     # assert xe.shape == (n1, n2)
     # assert np.array_equal(xe, z1)
 
+# ===============================================================================
+@pytest.mark.parametrize('dtype', [float, complex])
+@pytest.mark.parametrize('n1', [10, 17])
+@pytest.mark.parametrize('n2', [13, 7])
+@pytest.mark.parametrize('p1', [1, 2])
+@pytest.mark.parametrize('p2', [1])
+@pytest.mark.parametrize('s1', [1, 2])
+@pytest.mark.parametrize('s2', [1])
+@pytest.mark.parametrize('P1', [True, False])
+@pytest.mark.parametrize('P2', [True])
+@pytest.mark.parallel
+def test_stencil_vector_2d_parallel_array_to_psydac(dtype, n1, n2, p1, p2, s1, s2, P1, P2):
+    npts = [n1, n2]   
+
+    from mpi4py import MPI
+    comm = MPI.COMM_WORLD
+
+    # Create domain decomposition
+    D = DomainDecomposition(npts, periods=[P1, P2], comm=comm)
+
+    # Partition the points
+    global_starts, global_ends = compute_global_starts_ends(D, npts)
+    C = CartDecomposition(D, npts, global_starts, global_ends, pads=[p1, p2], shifts=[s1, s2])
+
+    # Create vector space and stencil vector
+    V = StencilVectorSpace(C, dtype=dtype)
+    x = StencilVector(V)
+
+    # Fill the vector with data
+
+    if dtype == complex:
+        f = lambda i1, i2: 10j * i1 + i2
+    else:
+        f = lambda i1, i2: 10 * i1 + i2
+    for i1 in range(V.starts[0], V.ends[0]+1):
+        for i2 in range(V.starts[1], V.ends[1]+1):
+            x[i1, i2] = f(i1, i2)
+
+    # Convert vector to array
+    xa = x.toarray()
+
+    # Convert array to vector of V
+    v = array_to_psydac(xa, V)
+
+    # Test properties of v and data contained
+    assert isinstance(v, StencilVector)
+    assert v.space is V
+    assert np.array_equal(xa, v.toarray())
```

Comment:
Yes indeed! This will also run faster:
```suggestion
    assert np.array_equal(x._data, v._data)
```

Reply:
One problem is that by using `_data` we are also comparing the ghost regions. I believe that the ghost regions of `v` are up to date, but those of `x` are not. So an additional instruction should be added above:
```python
x.update_ghost_regions()
```

Reply:
But why remove the check that `xa` == `v.toarray()` ?
This makes sense to check that `toarray` is the (left) inverse of `array_to_psydac`.

Reply:
If there is a problem in the ghost regions the assert:
`assert np.array_equal(x._data, v._data)`
may fail, however `array_to_psydac` could still be the inverse of `toarray`. Since this test tests the later, I would keep:
`assert np.array_equal(xa, v.toarray())`

As an addition, we can also include the other assert.

Reply:
> But why remove the check that `xa` == `v.toarray()` ? This makes sense to check that `toarray` is the (left) inverse of `array_to_psydac`.

`v` is built from `x.toarray()`, if you go back to `x` there is nothing left to check.

Reply:
It seems to me that to verify that `array_to_psydac` is the inverse of `toarray` we should check that:
- it is a left inverse by comparing `x` and `v`
- it is a right inverse by comparing `xa` and `v.toarray()`
no ?

Reply:
I believe that if you want to check that `array_to_psydac` is a right inverse, then you should create `xa` arbitrarily and not define it as `xa = x.toarray()`.

Reply:
With the current definition for `xa`, `assert np.array_equal(xa, v.toarray())` is redundant.

Reply:
Done in [dc00ace](https://github.com/pyccel/psydac/pull/404/commits/dc00ace1e8981d08174713f664a80960bf2516b8)